kani-compiler/src/codegen_boogie/context/boogie_ctx.rs

```diff
@@ -13,14 +29,320 @@ pub struct BoogieCtx<'tcx> {
     /// a snapshot of the query values. The queries shouldn't change at this point,
     /// so we just keep a copy.
     pub queries: QueryDb,
+    /// the Boogie program
+    program: BoogieProgram,
+    pub hooks: BoogieHooks<'tcx>,
 }
 
 impl<'tcx> BoogieCtx<'tcx> {
     pub fn new(tcx: TyCtxt<'tcx>, queries: QueryDb) -> BoogieCtx<'tcx> {
-        BoogieCtx { tcx, queries }
+        BoogieCtx { tcx, queries, program: BoogieProgram::new(), hooks: fn_hooks() }
+    }
+
+    /// Codegen a function into a Boogie procedure.
+    /// Returns `None` if the function is a hook.
+    pub fn codegen_function(&self, instance: Instance<'tcx>) -> Option<Procedure> {
+        debug!(?instance, "boogie_codegen_function");
+        if self.hooks.hook_applies(self.tcx, instance).is_some() {
+            debug!("skipping hook function `{instance}`");
+            return None;
+        }
+        let mut decl = self.codegen_declare_variables(instance);
+        let body = self.codegen_body(instance);
+        decl.push(body);
+        Some(Procedure::new(
+            self.tcx.symbol_name(instance).name.to_string(),
+            vec![],
+            vec![],
+            None,
+            Stmt::Block { statements: decl },
+        ))
+    }
+
+    pub fn codegen_declare_variables(&self, instance: Instance<'tcx>) -> Vec<Stmt> {
```

Comment:
```suggestion
    fn codegen_declare_variables(&self, instance: Instance<'tcx>) -> Vec<Stmt> {
```

Reply:
Removed `pub` from this function and a bunch of other functions as well.